### ISSUE TITLE
FeatureRequest: modular extension API

### ISSUE THREAD

**Issue by telamon** [mafintosh/hypercore-protocol#46]
**TL;DR** Design challenge that prevents multiple extensions to co-exist on 1 channel without external `onextension` multiplexer.

Superseeds #27

Changes introduced by V7:
- Extensions can now be registered at any time while the stream is alive
- The event emitter was replaced with an extension handler, which solves potential issues with dangling listeners and accidental double handling of messages.

Since the event emitter is now gone, there is sadly only 1 non-extendable extension handler per channel.

**Recap from irc:**

> &lt;mafintosh&gt; we prob need a .register(name, handleObject) thing
> &lt;telamohn&gt; mafintosh: yeah that would be nice, but i don't want to overcomplicate the protocol either.
> &lt;telamohn&gt; or rather, _I_ don't want to be responsible for introducing more complexity xD
> &lt;mafintosh&gt; maybe just ext = feed.registerExtension(name, { onmessage })
> &lt;mafintosh&gt; ext.send(message)
> &lt;mafintosh&gt; that it can send the options message in the background etc
> &lt;mafintosh&gt; then it could take an encoding also

**Mock prototype 1**

A selfcontained extension that never sees the channel.
Honestly i get a bad gut feeling about this, not entirely sure why.

```js
// implementation.js
const extension = require('my-hypercore-protocol-extension')

const stream = new Protocol(true)
const channel = stream.open(key)

channel.register('my_extension_name', extension)
```

```js
// my-hypercore-protocol-extension/index.js
const codecs = require('codecs')

module.exports = {
  encoding: codecs('string'), // Optional encoding def

  // optional hook to let extension bootstrap
  onmounted (sendFn, registeredName) {
    // * sendFn: channel.extension.bind(channel, registeredName)
    // * registeredName: 'my_extension_name'

    this.send = sendFn // store sendFn ref for later.
  }

  // required hook
  onmessage (stringMsg) {
    console.log('String received and decoded', stringMsg)
    this.send('Hello!') // response sent
  }
}
```

**Mock prototype 2**
A standalone extension that is allowed
access to the channel.

```js
// implementation.js
const MyExtension = require('my-hypercore-protocol-extension')

const stream = new Protocol(true)
const channel = stream.open(key)

const myExt = new MyExtension(channel) // pass the channel.
```

```js
// my-hypercore-protocol-extension/index.js
const codecs = require('codecs')
const EXT_IDENTIFIER = 'some_ext'
class MyExtension {

  constructor (channel) {
    this.channel = channel  // cyclic ref
    channel.register(EXT_IDENTIFIER, this) // leak self.
  }

  onmessage (stringMsg) {
    console.log('String received and decoded', stringMsg)
    this.channel.extension(EXT_IDENTIFIER, 'Hello!') // response sent
  }
}
```

This issue has been on my mind for a long time now, but I haven't been able to make an API suggestion that doesn't itch. Please help

**Comment by mafintosh** [comment]
My gist for this looks like this (hypercore focused):

```js
const ext = feed.registerExtension('hello', {
  encoding: <set optional encoding for the message>,
  onmessage (decodedMessage, peer) {
    // ... do whatever you want
  }
})

ext.name // hello
ext.broadcast(message) // sends to every peer
ext.send(message, peer) // send to a specific peer

ext.destroy() // unregisters the extension
```

**Comment by telamon** [comment]
@mafintosh That's a nice interface apologies if i missed it, Is it already implemented in hypercore 
or it's being considered for implementation?

I highly value an interface that let's me write extensions compatible with vanilla hypercore and standalone replication-managers; given that they can provide the same `registerExtension()` behaviour of course.
(in my case [decentstack](https://github.com/decentstack/decentstack) )

I'm gonna try and port [substream-extension](https://github.com/telamon/hypercore-protocol-substream/) to use the pattern you described, will reply here if i hit a snag.
Thanks again!

**Comment by mafintosh** [comment]
Not implemented yet, but I'll look like that when I get to it. If someone wants to beat me to it feel free :)

**Comment by telamon** [comment]
@mafintosh um how will `.registerExtension()` affect hypercore's `opts.extension` prop? 
Convert it to a an object? `{ extName: extObj }`

I wonder if it should be deprecated..
 In most code I've come across people are diligently fowarding that option but I can't recall anyone actually using it.

**Comment by mafintosh** [comment]
Yea we’ll deprecate it

**Comment by telamon** [comment]
Just realized that it was implemented in hypercore almost a month ago.
https://github.com/mafintosh/hypercore/pull/231
Great work!
Closing this issue.
